hal/spi_api.h

```diff
@@ -19,220 +18,345 @@
+    bool clock_phase; /**< True if data line is valid when leaving active state. */
+    bool clock_polarity; /**< True if the clock's rest state is high (+Vcc). */
+    bool continuous_mode; /**< True to use the continuous mode. */
+
+    spi_mode_t  mode; /**< Transmission mode. See spi_mode_t. */
+} spi_init_t;
 
 /**
- * \defgroup hal_GeneralSPI SPI Configuration Functions
- * @{
+ * This enumerates the possible result of the spi_init function.
  */
+typedef enum {
+    SPI_RESULT_OK, /**< Operation successful. */
+    SPI_RESULT_CONFIG_UNSUPPORTED, /**< The required parameters are not supported by the device. */
+    SPI_RESULT_INVALID_PARAM, /**< The given parameter(s) is/are invalid. */
+    SPI_RESULT_DEVICE_BUSY, /**< The requested peripheral/SS p)in is already initialised. */
+} spi_result_t;
 
-/** Initialize the SPI peripheral
+/**
+ * This is describes the capabilities of a SPI channel.
+ * The elements of this structure are independent. This means that a supported feature is supported
+ * regardless of other feature settings.
  *
- * Configures the pins used by SPI, sets a default format and frequency, and enables the peripheral
- * @param[out] obj  The SPI object to initialize
- * @param[in]  mosi The pin to use for MOSI
- * @param[in]  miso The pin to use for MISO
- * @param[in]  sclk The pin to use for SCLK
- * @param[in]  ssel The pin to use for SSEL
+ * @note minimum_frequency and maximum_frequency are used by tests to assess that at least these
+ * frequencies are supported. The device may be able to achieve lower and/or faster frequency.
+ * @warning Only minimum_frequency and maximum_frequency are assessed during tests. Your device may
+ * not be able to generate the whole range between these limits.
  */
-void spi_init(spi_t *obj, PinName mosi, PinName miso, PinName sclk, PinName ssel);
+typedef struct {
+    uint32_t    minimum_frequency; /**< Minimum frequency assessed during tests. */
+    uint32_t    maximum_frequency; /**< Maximum frequency assessed during tests. */
+
+    uint32_t    word_length; /**< Each bit represents the corresponding word length.
+                                  lsb => 1bit, msb => 32bit.*/
+    enum {
+        SPI_CONTINUOUS_MODE_NONCONTINUOUS, /**< The channel only supports the noncontinuous mode. */
+        SPI_CONTINUOUS_MODE_CONTINUOUS, /**< The channel only supports the continuous mode. */
+        SPI_CONTINUOUS_MODE_BOTH, /**< The channel supports both mode. */
+    } continuous_mode;
+    enum {
+        SPI_AUTHORITY_MASTER, /**< The device only supports the master mode. */
+        SPI_AUTHORITY_SLAVE, /**< The device only supports the slave mode. */
+        SPI_AUTHORITY_BOTH, /**< The device supports both modes. */
+    } authority; /**< Describes the bus authority capability of this channel. */
+    bool        support_half_duplex; /**< The device supports the half duplex (aka 3 wires) mode. */
+    bool        support_dual_io; /**< The device supports the 2 wires mode. */
```

Comment:
> + * - The hal implementation must support the simplex and full duplex modes

Does the above statement mean dual_io is must?

Reply:
No, `full duplex` uses MISO and MOSI while `dual-io` is a half-duplex mode.
`full-duplex` mode is required, `dual-io` is optional.

Reply:
I think these bools need clarification.
- `support_half_duplex`
- `support_dual_io`
- `support_quad_io`

My quick survey suggests there is a variety of nomenclature around (e.g. I've seen the one with SSEL, SCK, and SDIO referred to as both 2-wire and 3-wire), so I'd suggest naming the signals used by each mode, in addition to whatever other necessary verbiage.